### ISSUE TITLE
Unable to update implicitly to Microsoft.AspNetCore.App

### ISSUE THREAD

**Issue by benrhere** [aspnet/Home#3561]
Create a brand new ASP.NET Core, empty project in VS 15.8.5. My machine ONLY has SDK 2.1.402 installed:

![image](https://user-images.githubusercontent.com/1451620/46152197-86323b80-c23e-11e8-856e-ab58e873ee1d.png)

This documentation : https://docs.microsoft.com/en-us/aspnet/core/fundamentals/metapackage-app?view=aspnetcore-2.1

states:

To update ASP.NET Core:
On development machines and build servers: Download and install the .NET Core SDK.
On deployment servers: Download and install the .NET Core runtime.
Applications will roll forward to the latest installed version on application restart. It's not necessary to update the Microsoft.AspNetCore.App version number in the project file. 

Yet, this is not happening. Restart, and you still will see:

![image](https://user-images.githubusercontent.com/1451620/46152476-13759000-c23f-11e8-9077-78384e7fc5ab.png)

The unmodified project file looks like:

```
<Project Sdk="Microsoft.NET.Sdk.Web">
​
  <PropertyGroup>​
    <TargetFramework>netcoreapp2.1</TargetFramework>​
  </PropertyGroup>​
​
  <ItemGroup>​
    <Folder Include="wwwroot\" />​
  </ItemGroup>​
​
  <ItemGroup>​
    <PackageReference Include="Microsoft.AspNetCore.App" />​
  </ItemGroup>​
​
</Project>​

```
If I specify an explicit version on the packagereference above, it works. But that contradicts the documentation. Based on some searches, there is a lot of confusion around this. The NuGet UI "learn more" link is unhelpful, just pointing to info about updating NuGet packages in general.

I haven't even begun to investigate the situation with Microsoft.NETCore.App displaying similar behavior. This feels like a rough developer experience.

**Comment by natemcmaster** [comment]
Our documentation on this package was recently yesterday to add more details about why you should not specify an explicit version. See https://docs.microsoft.com/en-us/aspnet/core/fundamentals/metapackage-app?view=aspnetcore-2.1. I also wrote a blog post explaining even more details on this: https://natemcmaster.com/blog/2018/08/29/netcore-primitives-2/#relationship-to-the-nuget-package.

**TLDR;** The short version is that the package version represents the **minimum** version of the ASP.NET Core shared framework. You don't need to update the package reference for application rollforward to happen.

FYI - you're not the first to ask this question. Hopefully this helps: https://github.com/NuGet/docs.microsoft.com-nuget/pull/1067

Let me know if the details help.

**Comment by benrhere** [comment]
Hi @natemcmaster and @Eilon , thanks for your response. I did read through the document that you referenced before posting the question.

Are you saying that the version displayed there in NuGet has nothing to do with the version which will actually be pulled when I hit F5? (The version it points to isn't even installed on my machine.) Also, when I explicitly update the reference in the csproj file to point to 2.1.4, it does reflect that in the UI and seems to repull nuget packages.

If I'm understanding you correctly, it feels strange that NuGet is not showing me the version that I'll consume, and it's stating that I'm "blocked" implying that I'm stuck on 2.1.1 when I'm not even going to use that version. Ultimately, I have no way of seeing what version I'm pointing at it seems.

**Comment by natemcmaster** [comment]
My blog post walks through the details of what is happening. I recommend giving this a read too. https://natemcmaster.com/blog/2018/08/29/netcore-primitives-2/

To be clear, what happens when you press F5 is:

1. Your project has a NuGet package reference to Microsoft.AspNetCore.App 2.1.1 (the SDK picked 2.1.1, not your project)
2. NuGet restore packages into $HOME/.nuget/packages
3. Visual Studio builds your project. One of the build outputs is `bin/Debug/netcoreapp2.1/MyApp.runtimeconfig.json`, which looks like:
```json
{
  "runtimeOptions": {
    "framework": {
      "name": "Microsoft.AspNetCore.App",
      "version": "2.1.1"
    }
  }
}
```
4. Visual Studio launches your project by running `C:\Program Files\dotnet\dotnet.exe bin/Debug/netcoreapp2.1/MyApp.dll`. While loading your app, it will look to see what is in `C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App\`. It may be that you have `C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App\2.1.1`, but if there is  `C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App\2.1.4` or `C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App\2.1.9999`, it's going to roll-forward and pick the highest 2.1.* version based on the rules described in https://natemcmaster.com/blog/2018/08/29/netcore-primitives-2/#version-roll-forward

Please note that in step 4 (when your app loads) the files in $HOME/.nuget/packages are not part of the equation. Your app only uses the NuGet package as part of *building* the app, not *running* the app.

For what it's worth, we realize this is a confusing way for NuGet to behave, so we are working on changes in .NET Core 3.0 and Visual Studio vNext to make this more intuitive.

**Comment by benrhere** [comment]
Thanks for the clarification! I apologize that I missed your blog post. I'll spend time reading it and your above response to make more sense of this.
